File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.java
class Solution {
    public int subarraySum(int[] nums, int k) {
        Map<Integer, Integer> map = new HashMap<>();
        map.put(0, 1);

        int curSum = 0;
        int cnt = 0;

        for (int i = 0; i < nums.length; i++) {
            curSum += nums[i];
            
            if (map.containsKey(curSum - k)) {
                cnt += map.get(curSum - k);
            }

            map.put(curSum, map.getOrDefault(curSum, 0) + 1);
        }

        return cnt;
    }
}

File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.kt
class Solution {
    fun subarraySum(nums: IntArray, k: Int): Int {
        val map = mutableMapOf(0 to 1)
        var curSum = 0
        var cnt = 0

        for (num in nums) {
            curSum += num
            cnt += map.getOrDefault(curSum - k, 0)
            map[curSum] = map.getOrDefault(curSum, 0) + 1
        }

        return cnt
    }
}